src/Build/Graph/ProjectGraph.cs

```diff
@@ -185,28 +185,52 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
                 entryPointConfigurationMetadata.Add(configurationMetadata);
             }
 
-            LoadGraph(projectsToEvaluate, projectCollection, tasksInProgress);
-            foreach(var configurationMetadata in entryPointConfigurationMetadata)
+            bool exceptionInLoading = false;
+            if (LoadGraph(projectsToEvaluate, projectCollection, tasksInProgress, out List<Exception> exceptions))
             {
-                entryPointNodes.Add(_allParsedProjects[configurationMetadata]);
-                if (!nodeStates.TryGetValue(_allParsedProjects[configurationMetadata], out var _))
+                foreach (var configurationMetadata in entryPointConfigurationMetadata)
                 {
-                    DetectCycles(_allParsedProjects[configurationMetadata], nodeStates, projectCollection, configurationMetadata.GlobalProperties);
+                    entryPointNodes.Add(_allParsedProjects[configurationMetadata]);
+                    if (!nodeStates.TryGetValue(_allParsedProjects[configurationMetadata], out var _))
+                    {
+                        try
+                        {
+                            DetectCycles(_allParsedProjects[configurationMetadata], nodeStates, projectCollection, configurationMetadata.GlobalProperties);
+                        }
+                        catch (CircularDependencyException ex)
+                        {
+                            if (exceptions == null)
+                            {
+                                exceptions = new List<Exception>();
+                            }
+                            exceptions.Add(ex);
+                            exceptionInLoading = true;
+                        }
+                    }
                 }
-            }
 
-            var graphRoots = new List<ProjectGraphNode>(entryPointNodes.Count);
-            foreach(var entryPointNode in entryPointNodes)
-            {
-                if (entryPointNode.ReferencingProjects.Count == 0)
+                var graphRoots = new List<ProjectGraphNode>(entryPointNodes.Count);
+                foreach (var entryPointNode in entryPointNodes)
                 {
-                    graphRoots.Add(entryPointNode);
+                    if (entryPointNode.ReferencingProjects.Count == 0)
+                    {
+                        graphRoots.Add(entryPointNode);
+                    }
                 }
+
+                EntryPointNodes = entryPointNodes.AsReadOnly();
+                ProjectNodes = _allParsedProjects.Values.ToList();
+                GraphRoots = graphRoots.AsReadOnly();
+            }
+            else
+            {
+                exceptionInLoading = true;
```

Comment:
I think `exceptionsInLoading` can be replaced with `exceptions.Count > 0`

Reply:
Or, since `LoadGraph` only returns false when there are exceptions, jsut directly throw here instead of below.